### PR TITLE
fix(dns): Eliminate asynchronous timer in syncQuery() to prevent hang risk

### PR DESCRIPTION
**Pull request by chobits**
### Summary

Originally the first request to syncQuery() will trigger an asynchronous timer event, which added the risk of the thread pool hanging in the timerng library. With this patch, it will perform a synchronously DNS query for the first time if current phase is yieldable. 

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* syncQuery() will perform a synchronously DNS query for the first time if current phase is yieldable.
  * Keep the original async query using event timer for some unyieldable phase like init_worker_by_lua.
* simplify the logic of yieldable checking by using kong.tools.yield.in_yieldable_phase()
* simplify executeQuery() by using individualQuery()
* revert #11869

### Issue reference

Fix KAG-2913 FTI-5348

### CODE REVIEW

File: kong/resty/dns/client.lua [hunk 1]
```diff
@@ -839,7 +813,37 @@ local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
   local item = queue[key]
 
-  -- if nothing is in progress, we start a new async query
+  local is_yieldable = in_yieldable_phase()
+
+  -- If nothing is in progress and in the yieldable phase,
+  -- we start a new sync query
+
+  if not item and is_yieldable then
```
Comment (oowl) [inline kong/resty/dns/client.lua:821]: So if we want to query DNS in the `init_worker` phase, the logic can not query any DNS name, Is this a logic bug for Kong,. In our old logic, we can trigger DNS query in any openresty phase, I think we need to think about this carefully.

Comment (chobits) [reply kong/resty/dns/client.lua:821]: yeah, if it is in init_worker phase, it will run the following logic which is kept by this pr. See init_worker phase's running logic: https://github.com/Kong/kong/pull/11900/files#diff-2001be5a8987462f9edc79d36a2f3c3a7138ae56a2aca1ff75573d19127edf7eR847( also for other unyieldable phases)

File: kong/resty/dns/client.lua [hunk 2]
```diff
@@ -839,7 +813,37 @@ local function syncQuery(qname, r_opts, try_list)
...
+
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
+
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
+    queue[key] = nil
+    -- 2) release all waiting threads
+    item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
+    item.semaphore = nil
+    ngx.sleep(0)
```
Comment (chronolaw) [inline kong/resty/dns/client.lua:839]: Should we use `require("kong.tools.yield").yield()`?

File: kong/resty/dns/client.lua [hunk 3]
```diff
@@ -839,7 +813,37 @@ local function syncQuery(qname, r_opts, try_list)
...
+    queue[key] = item
+
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
+
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
+    queue[key] = nil
+    -- 2) release all waiting threads
+    item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
+    item.semaphore = nil
```
Comment (fffonion) [inline kong/resty/dns/client.lua:838]: I noticed this is copied from `executeQuery`, but it seems we aren't check item.semaphore null-ness before using it on Line 870. Should we add a check?

File: kong/resty/dns/client.lua (hunk 2 above)
Comment (chobits) [reply kong/resty/dns/client.lua:839]: yeah, good catch, using `yield` api is better, will update later

File: kong/resty/dns/client.lua (hunk 3 above)
Comment (chobits) [reply kong/resty/dns/client.lua:838]: let me check, taking a look into it

File: kong/resty/dns/client.lua (hunk 2 above)
Comment (chobits) [reply kong/resty/dns/client.lua]: fixed

File: kong/resty/dns/client.lua (hunk 3 above)
Comment (chobits) [reply kong/resty/dns/client.lua:839]: For the completion, we need to check its null-ness. But the semaphore.new() is very based resource allocation function, i check its implmentation, there is only one error situation where the memory allocation fails. So I think checking its nullness causes code redudancy, which is like checkting lua table new allocation. And in kong source base, most calls of semaphore.new() does not check its return value. And in its official documentation, it does not check its return value, see [here](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#synchronizing-threads-in-the-same-context)

File: changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml [hunk 4]
```diff
@@ -1,3 +0,0 @@
-message: The default value of `dns_no_sync` option has been changed to `on`
```
Comment (vm-001) [inline changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml:1]: Is this PR going to be released at 3.5? I'm wondering why this file is being deleted.

File: kong/resty/dns/client.lua [hunk 5]
```diff
@@ -747,41 +748,15 @@ local queue = setmetatable({}, {__mode = "v"})
...
   -- 1) stop new ones from adding to our lock/semaphore
   queue[item.key] = nil
   -- 2) release all waiting threads
   item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
   item.semaphore = nil
-  ngx.sleep(0)
-  -- 3) destroy the resolver -- ditto in individualQuery
-  if r then
-    r:destroy()
-  end
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:784]: After the change the `destroy()` happens at a different time (before the yield). Do we have some reason to destroy after the yield in the original code?

File: kong/resty/dns/client.lua [hunk 6]
```diff
@@ -747,41 +748,15 @@ local queue = setmetatable({}, {__mode = "v"})
...
+  yield()
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:759]: What is this `yield()` for?

File: changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml (hunk 4 above)
Comment (chobits) [reply changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml:1]: yes. This patch try to fix dns blocking issue, so we need to revert original temporary fix that changes  `dyns_no_sycn` option to `on`

Comment (chobits) [reply changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml:1]: Sry, my mistake, this patch will not be merged into 3.5.0.0, because it's already frozen.
But it could be involved in 3.5.x.x.

Comment (vm-001) [reply changelog/unreleased/kong/fix_dns_enable_dns_no_sync.yml:1]: If you're trying to ship it in patch release of 3.5.0, please restore this changelog file.

File: kong/resty/dns/client.lua [hunk 7]
```diff
@@ -839,7 +814,37 @@ local function syncQuery(qname, r_opts, try_list)
...
+
+  if not item and is_yieldable then
+    item = {
+      key = key,
+      semaphore = semaphore(),
+      qname = qname,
+      r_opts = cycle_aware_deep_copy(r_opts),
+      try_list = try_list,
+    }
+    queue[key] = item
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:830]: I presume this is also done in `asyncQuery()`. In that case, we'd better extract the part that handles queue and semaphore and not copy&paste.

Comment (StarlightIbuki) [reply kong/resty/dns/client.lua:830]: A possible name for the extracted function: "`acquire_lock_query`" and "`release_lock_query`".

File: kong/resty/dns/client.lua (hunk 6 above)
Comment (chobits) [reply kong/resty/dns/client.lua:759]: I examined the original calls to `ngx.sleep(0)`/`yield()` and found that `yield` call has no practical meaning. The original logic calls `ngx.sleep(0)` at the end of the timer executeQuery(), it implies the coroutine yield upon the timer completion.

File: kong/resty/dns/client.lua [hunk 8]
```diff
@@ -839,7 +812,36 @@ local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
   local item = queue[key]
 
-  -- if nothing is in progress, we start a new async query
+  local is_yieldable = in_yieldable_phase()
+
+  -- If nothing is in progress and in the yieldable phase,
+  -- we start a new sync query
+
+  if not item and is_yieldable then
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:820]: I'd like not to do this:

```lua
if not item and is_yieldable then
...
end

if not item then
...
end


```

instead, I prefer:

```lua
if not item then
  if is_yieldable then
  ...
  else
end
```

And with further refactoring, with a reversed condition, we could do this:

```lua
if item then
  ...
  return ...
end

if is_yieldable then
  ...
else
 ...
end
```

At least, let's not check the not item condition twice.

File: kong/resty/dns/client.lua [hunk 9]
```diff
@@ -838,39 +811,45 @@ end
...
+  -- If not yieldable, we start a new async query and return nil
+  if not is_yieldable then
     if not item then
-      return item, err, try_list
+      asyncQuery(qname, r_opts, try_list)
     end
-  else
-    add_status_to_try_list(try_list, "in progress (sync)")
+    -- item.result is definitely nil in this situation
+    return nil, "API disabled in the unyieldable phase", try_list
```
Comment (chobits) [inline kong/resty/dns/client.lua:822]: TIPS: [The original logic](https://github.com/Kong/lua-resty-dns-client/pull/130/files#diff-c220f0318ab8cbd55a2893ac929b5b0337a36c8046eb52d2e11e9880f985583bR842) returns `items + nil`, it breaks the convention of syncQuery() API, in its comment, we know that it only returns `item.record, nil` or `nil, err`.

But this wrong return value `item` can be handled by [its caller `resolver()` ](https://github.com/Kong/kong/blob/master/kong/resty/dns/client.lua#L1251), however, `syncQuery()` is not exported from this library, so the user of this library will not notice this wrong logic.

File: kong/resty/dns/client.lua (hunk 3 above)
Comment (chobits) [reply kong/resty/dns/client.lua:844]: hi @fffonion I have reviewed the logic, thinking it is unnecessary to check for nullness. Except for the scenario where it can between obtaining the `item` and using the `item.semaphore`, I have checked all instances of this type of context, and I couldnt find any explicit `yield()` or `ngx.sleep`/network-IO operation.

File: kong/resty/dns/client.lua (hunk 5 above)
Comment (chobits) [reply kong/resty/dns/client.lua:784]: carefully examined the logic and referred to the official resolver library's examples. Here, we can proactively destroy 'r' since we've already extracted the data from it.

File: kong/resty/dns/client.lua (hunk 7 above)
Comment (chobits) [reply kong/resty/dns/client.lua:835]: i agree, but refactoring the original logic is not the goal of this PR. And the task of improving this library is in the plan.

File: kong/resty/dns/client.lua (hunk 8 above)
Comment (chobits) [reply kong/resty/dns/client.lua]: resolved

**Comment by chobits** [comment]
Some tips for reviewers:

### how this patch works in syncQuery()

Because the `git diff` makes the `file changed` uneasy to review. I describe my modification here:

the original logic of syncQuery():

```
if not item then
   asyncQuery()
   -- pass through to wait semaphore
else
    add_status_to_try_list(try_list, "in progress (sync)")
end

if not supported_semaphore_wait_phases then
  return item, nil, try_list.   -- here is a bug, we could not return item directly. (1)
end

semaphore:wait()
```

the modified logic of syncQuery():

```
if not is_yieldable then
  if not item then
     asyncQuery()
  end
  return nil, "API disabled in the unyieldable phase", try_list.  -- try to fix the bug (1)
end

if not item then
   item.result, item.err = individualQuery()    -- real synchronous query
   return item.result, item.err, try_list
end

sema:wait()
```

File: kong/resty/dns/client.lua (hunk 9 above)
Comment (StarlightIbuki) [reply kong/resty/dns/client.lua:822]: "item" is essentially a future object here. IMO, the library's interface should be refactored for all uses "future" and users may have the right expectation about how it could be used (after all future does not look like working with init phase).

File: kong/resty/dns/client.lua [hunk 10]
```diff
@@ -838,39 +811,45 @@ end
...
 
-  -- if nothing is in progress, we start a new async query
-  if not item then
-    local err
-    item, err = asyncQuery(qname, r_opts, try_list)
+  -- If not yieldable, we start a new async query and return nil
+  if not is_yieldable then
     if not item then
-      return item, err, try_list
+      asyncQuery(qname, r_opts, try_list)
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:819]: This makes no sense. The call is rejected and returning error, but yet we initiate a query,

File: kong/resty/dns/client.lua [hunk 11]
```diff
@@ -838,39 +811,45 @@ end
...
-  if not supported_semaphore_wait_phases[ngx_phase] then
-    -- phase not supported by `semaphore:wait`
-    -- return existing query (item)
-    --
-    -- this will avoid:
-    -- "dns lookup pool exceeded retries" (second try and subsequent retries)
-    -- "API disabled in the context of init_worker_by_lua" (first try)
-    return item, nil, try_list
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
```
Comment (StarlightIbuki) [inline kong/resty/dns/client.lua:840]: We should avoid copy&pasting.
At least this should be extracted into a function.

Comment (StarlightIbuki) [reply kong/resty/dns/client.lua:840]: And a possible way to refactor it:
Introduce a type named future, and a future object is returned by the call.
Caller calls future:get() to wait for the value (in which it waits for the semaphore and value);
Callee calls future:set() to post the semaphore and return the value.

File: kong/resty/dns/client.lua (hunk 10 above)
Comment (chobits) [reply kong/resty/dns/client.lua]: yes, we will delete it in this pr. done

File: kong/resty/dns/client.lua (hunk 11 above)
Comment (chobits) [reply kong/resty/dns/client.lua]: done , abstract common function for item creating and releasing

File: kong/resty/dns/client.lua (hunk 9 above)
Comment (chobits) [reply kong/resty/dns/client.lua]: OK I will add this improvement to our future ticket to refactor this lib

**Review by StarlightIbuki — APPROVED** [review]
Approve given the CI is green.

**Review by fffonion — APPROVED** [review]

**Review by oowl — APPROVED** [review]
LGTM

**Comment by chobits** [comment]
wait for me to fix test cases before merging

File: kong/resty/dns/client.lua [hunk 12]
```diff
@@ -742,46 +741,38 @@ local function individualQuery(qname, r_opts, try_list)
...
-      parseAnswer(item.qname, item.r_opts.qtype, item.result, item.try_list)
-      --[[
-      log(DEBUG, PREFIX, "Query parsed answer: ", item.qname, ":", item.r_opts.qtype, " ", fquery(item),
-              " ", frecord(item.result))
-    else
-      log(DEBUG, PREFIX, "Query error: ", item.qname, ":", item.r_opts.qtype, " err=", tostring(err))
-      --]]
-    end
-  end
+local function queue_get_item(key, qname, r_opts, try_list)
```
Comment (hanshuebner) [inline kong/resty/dns/client.lua:745]: Move the `timer_at` invocation into this function and rename it `enqueue_query`.  Also rename `item` to `query`.

File: kong/resty/dns/client.lua [hunk 13]
```diff
@@ -742,46 +741,38 @@ local function individualQuery(qname, r_opts, try_list)
...
+    qname = qname,
+    r_opts = cycle_aware_deep_copy(r_opts),
+    try_list = try_list,
+  }
+  queue[key] = item
+  return item
+end
 
+
+local function queue_release_item(item)
```
Comment (hanshuebner) [inline kong/resty/dns/client.lua:758]: Rename this function to `release_requests_waiting_for_query`.

**Review by hanshuebner — CHANGES_REQUESTED** [review]
I like the extraction into named functions, but the naming can be improved.

**Review by Tieske — COMMENTED** [review]
I do not think this PR will solve the underlying issue. It might reduce probability of it occurring though.

The old logic would always use a timer to to the query. In case of a `syncQuery`, it would immediately after scheduling the query block on the semaphore and wait for it to be finished.

In the proposed code in this PR, a `syncQuery` will be done (without a timer now), and all other requests will wait for it. Which seems to solve the issue since no timers are involved.

However; when resolving a name finds the cached data to be stale, it will still do an `asyncQuery` which will involve a timer. If that timer doesn't execute, the exact same problem will still occur. All requests will keep blocking on the semaphore until it times out.

My guess is that most cases will hit the `asyncQuery` because of stale data. Only initially when the cache is empty we'll be doing `syncQuery` (and possibly later on on low-volume routes where the DNS cache entry expires its TTL+stale time before the next request arrives).

And I do not see any way around it. Making everything sync will have latency costs because of in-band lookups. Doing async can fail on timers. The only way would be to have more sturdy timers that will not fail.

TLDR: when timers start failing, all bets are off.

**Review by dndx — CHANGES_REQUESTED** [review]
This is not defensive enough, two additional cases must be considered:

If stale entry is detected inside `queue`, it must be removed and an error log printed. 

If semaphore timeouts, a sync query must be started immediately instead of returning an error from the semaphore.

We have to assume every part of the chain can go around, and design robust error handling that accommodates each case, no matter how small the chances are.

File: kong/resty/dns/client.lua (hunk 12 above)
Comment (chobits) [reply kong/resty/dns/client.lua:745]: > Move the timer_at invocation into this function 
The timer_at is not the common logic for this function. There are two scenarios of calling this function:
1. queue_get_item -> timer_at(asynchronous query)
2. queue_get_item -> individualQuery (synchronous query)

> rename it `enqueue_query`
No problem, I'll update it later

> Also rename `item` to `query`.

This will introduce many modifications. This pr is an urgent bugfix, we will redesign this module in kong-3.6, I believe it is a good opportunity to refactor this at that time.

**Comment by chobits** [comment]
TODO:

1. detect whether the query semaphore:wait is blocked for a long time, which maybe a bug. Do we need to check this?
2. Make sure if balancer(upper caller) has the retry mechanism when using this libraray, otherwise we need to retry it when resolving failed after it wakes up from sempaphore:wait()

**Comment by Tieske** [comment]
> > And I do not see any way around it. Making everything sync will have latency costs because of in-band lookups. Doing async can fail on timers.
> 
> No, this patch will not introduce new latency, because the original implementation of asyncTImer already introduced the latency that other requests would wait on it. that means

Sorry didn't mean to say the patch introduced it. Just in general switching everything to sync would introduce latency.

**Comment by Tieske** [comment]
@chobits I think that an expire-time should be included in the query-item. Upon creation as `now() + resolve_max_wait`. 

Every next call that waits for the semaphore, should set the timeout to `item.expire_time - now()` instead of just `resolve_max_wait` which it currently does. That ensures waiting threads will return asap.

Next thing would be to wrap checks like this;
```lua
  local item = queue[key]

  -- If nothing is in progress, we start a new sync query
  if not item then
```

Into its own function `get_query(key)`. This function can then check based on the expire-time if the thing in the queue is still valid, and if not take corrective action (remove from queue, and release semaphore).

**Comment by Tieske** [comment]
> If semaphore timeouts, a sync query must be started immediately instead of returning an error from the semaphore.

This is very dangerous imho. the reason timers do not execute in the first place is because the system is under stress. So why would a retry in this case solve the issue? quite likely the new try will fail for the same reason, since we're adding to the stress.

Consider; 200 rps, and `resolve_max_wait = 15`. So when we hit the first failure, we already have 15 * 200 requests waiting; 3000. Then we retry, fail again; 6000 will be waiting by then, and so on.

The system was already under stress causing the first failure, hoarding those requests will only increase the stress and reduce likelihood of recovery. Similar to a kernel killing random processes, we should drop the requests, as that is the best chance of survival.

**Comment by chobits** [comment]
> @chobits I think that an expire-time should be included in the query-item. Upon creation as `now() + resolve_max_wait`.
> 
> Every next call that waits for the semaphore, should set the timeout to `item.expire_time - now()` instead of just `resolve_max_wait` which it currently does. That ensures waiting threads will return asap.
> 

Agree, will fix it later.

> Next thing would be to wrap checks like this;
> 
> ```lua
>   local item = queue[key]
> 
>   -- If nothing is in progress, we start a new sync query
>   if not item then
> ```
> 
> Into its own function `get_query(key)`. This function can then check based on the expire-time if the thing in the queue is still valid, and if not take corrective action (remove from queue, and release semaphore).

Will fix it later, @dndx . This checking and releasing logic is what you recommend us to do to handle the situation when semaphore waiting queue is blocked.

**Comment by chobits** [comment]
hi, pls review the new introduced commit https://github.com/Kong/kong/pull/11900/commits/71b144cf18d8a64b50fccbbb01c395006416d1b0

I add bug-checking logic as follows:
1. Firstly, retrieve the query item. If it's found to be expired, proceed to release all waiting queries.
2. If the semaphore:wait() operation fails, report an alert-level error log.

File: kong/resty/dns/client.lua [hunk 14]
```diff
@@ -742,46 +744,59 @@ local function individualQuery(qname, r_opts, try_list)
...
+    -- 1) stop new ones from adding to our lock/semaphore
+    queue[item.key] = nil
+    -- 2) release all waiting threads
+    item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
+    item.semaphore = nil
+  end
+end
+
+
+local function queue_get_query(key)
```
Comment (Tieske) [inline kong/resty/dns/client.lua:774]: can we add the `try_list` to the parameters, and add to it when the query has expired?

Comment (chobits) [reply kong/resty/dns/client.lua]: done

File: t/03-dns-client/01-phases.t [hunk 15]
```diff
@@ -59,8 +59,7 @@ qq {
 GET /t
 --- response_body
 answers: nil
-err: dns client error: 101 empty record received
---- no_error_log
+err: nil
+--- error_log
```
Comment (Tieske) [inline t/03-dns-client/01-phases.t:63]: @chobits why did this test result change?

Comment (chobits) [reply t/03-dns-client/01-phases.t:63]: This test case was introduced by this [PR](https://github.com/Kong/lua-resty-dns-client/pull/130/files) , and its scenario will not be used in kong project because this library has been imported into kong, we do not run dns.client query in init_worker phase.

So I eliminate asynchronous timer in syncQuery() in this PR, after this, the test case result will change. Now if you call dns query API in init_worker phase, it will be disabled directly.

File: t/03-dns-client/02-timer-usage.t [hunk 16]
```diff
@@ -1,78 +0,0 @@
...
-plan tests => repeat_each() * (blocks() * 5);
-
-workers(6);
-
-no_shuffle();
-run_tests();
-
-__DATA__
-
-=== TEST 1: reuse timers for queries of same name, independent on # of workers
```
[the remark this answers is not in the document]
Comment (chobits) [reply t/03-dns-client/02-timer-usage.t:12]: Because `02-timer-usage.t` did not test the scenario when stale data triggerred the async timer. It only tested async query logic in syncQuery().

If we want to test stale data triggered async logic, we need to write a new one.

[the remark this answers is not in the document]
Comment (chobits) [reply t/03-dns-client/02-timer-usage.t:12]: hi @Tieske I added a new test case to check the scenario where the stale data triggers async timer

File: kong/resty/dns/client.lua [hunk 17]
```diff
@@ -651,7 +651,9 @@ _M.init = function(options)
 
   config = options -- store it in our module level global
 
-  resolve_max_wait = options.timeout / 1000 * options.retrans -- maximum time to wait for the dns resolver to hit its timeouts
+  -- maximum time to wait for the dns resolver to hit its timeouts
+  -- +1s to ensure semaphore:wait() expires after the DNS querying timeout
```
Comment (dndx) [inline kong/resty/dns/client.lua:655]:
```suggestion
  -- + 1s to ensure some delay in timer execution and semaphore return are accounted for
```

File: kong/resty/dns/client.lua [hunk 18]
```diff
@@ -742,46 +744,61 @@ local function individualQuery(qname, r_opts, try_list)
...
+local function queue_get_query(key, try_list)
+  local item = queue[key]
+
+  if not item then
+    return nil
+  end
+
+  -- bug checks: release it actively if the waiting query queue is blocked
+  if item.expire_time < time() then
+    local err = "query expired, key:" ..  key
```
Comment (dndx) [inline kong/resty/dns/client.lua:783]:
```suggestion
    local err = "stale query, key:" ..  key
```

File: kong/resty/dns/client.lua [hunk 19]
```diff
@@ -804,14 +821,7 @@ local function asyncQuery(qname, r_opts, try_list)
...
-    key = key,
-    semaphore = semaphore(),
-    qname = qname,
-    r_opts = cycle_aware_deep_copy(r_opts),
-    try_list = try_list,
-  }
-  queue[key] = item
+  item = enqueue_query(key, qname, r_opts, try_list)
 
   local ok, err = timer_at(0, executeQuery, item)
```
Comment (dndx) [inline kong/resty/dns/client.lua:826]: Why do we still create timers in here?

**Review by dndx — CHANGES_REQUESTED** [review]

File: kong/resty/dns/client.lua (hunk 19 above)
Comment (dndx) [reply kong/resty/dns/client.lua:826]: Answered offline.

**Review by ADD-SP — APPROVED** [review]

File: kong.conf.default [hunk 20]
```diff
@@ -1548,7 +1548,7 @@
 
 #dns_error_ttl = 1               # TTL in seconds for error responses.
 
-#dns_no_sync = on                # If enabled, then upon a cache-miss every
+#dns_no_sync = off               # If enabled, then upon a cache-miss every
```
Comment (chronolaw) [inline kong.conf.default:1551]: In PR #11869 we have turn on this flag, should we add another change log for it?

Comment (chobits) [reply kong.conf.default:1553]: I think we need to add this changed info to `[changelog/unreleased/kong/fix_dns_blocking.yml](https://github.com/Kong/kong/pull/11900/files#diff-dcaf6440867320b2760937851ee0fc8d15a37236ead390e382c9c956af88ee7d)`

```
message: Eliminate asynchronous timer in syncQuery() to prevent hang risk
```

->

```
message: Eliminate asynchronous timer in syncQuery() to prevent hang risk
                 The default value of `dns_no_sync` option has been changed to `off`
```

Comment (chobits) [reply kong.conf.default:1553]: hi @chronolaw See my new fix: https://github.com/Kong/kong/pull/11900/commits/d3ee61ce617694ee11da84742da48820f6053622. If this fix is ok, pls help to resolve this comment, ty!

Comment (dndx) [reply kong.conf.default:1553]: Please create two Changelog entries for these two changes.

Comment (chobits) [reply kong.conf.default:1553]: KK

Comment (chobits) [reply kong.conf.default:1553]: fixed with two changelog introduced

Comment (chronolaw) [reply kong.conf.default:1553]: great.

File: kong/resty/dns/client.lua [hunk 21]
```diff
@@ -837,42 +847,26 @@ end
...
+    return item.result, item.err, try_list
   end
 
+  -- If the query is already in progress, we wait for it.
+
+  add_status_to_try_list(try_list, "in progress (sync)")
+
   -- block and wait for the async query to complete
-  local ok, err = item.semaphore:wait(resolve_max_wait)
+  local ok, err = item.semaphore:wait(item.expire_time - time())
```
Comment (dndx) [inline kong/resty/dns/client.lua:869]: There is a risk that this might return negative value because of clock unwind. Suggest we keep `resolve_max_wait` as-is, it should be rare to wait for `resolve_max_wait` anyway. WDYT?

**Review by dndx — CHANGES_REQUESTED** [review]

File: kong/resty/dns/client.lua (hunk 21 above)
Comment (Tieske) [reply kong/resty/dns/client.lua:869]: > it should be rare to wait for resolve_max_wait anyway

yes, but at the same time; that's what this whole bug is about.

Thinking out loud; if we stick to `resolve_max_wait` then the requests will not be released all at once. But if they retry, they will all end up queueing on a single query again so there is no benefit to that, and if they don't retry but fail, then we better fail them as early as possible to reduce pressure on the system.

tldr; imho we should return early, and wrap it in a `max` call; `math.max(item.expire_time - time(), 0.001)`

Comment (chobits) [reply kong/resty/dns/client.lua:869]: HI @dndx ,dont worry, the previous logic of calling queue_get_query(), will make sure `item.expire_time - time()` is larger than 0.

Comment (chobits) [reply kong/resty/dns/client.lua:869]: but i agree to round it to the nearest 0.001 seconds

Comment (chobits) [reply kong/resty/dns/client.lua]: done

**Review by Tieske — APPROVED** [review]

**Review by dndx — APPROVED** [review]

File: changelog/unreleased/kong/fix_dns_disable_dns_no_sync.yml [hunk 22]
```diff
@@ -0,0 +1,3 @@
+message: The default value of `dns_no_sync` option has been changed to `off`
+type: bugfix
```
Comment (dndx) [inline changelog/unreleased/kong/fix_dns_disable_dns_no_sync.yml:2]:
```suggestion
type: feature
```

**Review by dndx — CHANGES_REQUESTED** [review]

File: kong/resty/dns/client.lua [hunk 23]
```diff
@@ -651,7 +651,9 @@ _M.init = function(options)
 
   config = options -- store it in our module level global
 
-  resolve_max_wait = options.timeout / 1000 * options.retrans -- maximum time to wait for the dns resolver to hit its timeouts
+  -- maximum time to wait for the dns resolver to hit its timeouts
+  -- + 1s to ensure some delay in timer execution and semaphore return are accounted for
+  resolve_max_wait = options.timeout / 1000 * options.retrans + 1
```
Comment (nowNick) [inline kong/resty/dns/client.lua:656]: What could happen if we didn't include this `+1` ?

File: spec/01-unit/21-dns-client/02-client_spec.lua [hunk 24]
```diff
@@ -1767,7 +1773,9 @@ describe("[DNS client]", function()
           touch = 0,
           expire = gettime() + 10,
         }
-        sleep(0.5) -- wait before we return the results
+        -- wait before we return the results
+        -- `+ 2` s ensures that the semaphore:wait() expires
+        sleep(timeout/1000 + 2)
```
[the remark this answers is not in the document]
Comment (nowNick) [reply spec/01-unit/21-dns-client/02-client_spec.lua:1778]: Also would it be possible to use some kind of waiter function pattern that would check against a predicate every _interval_ ?

**Review by nowNick — APPROVED** [review]
Just 2 concerns regarding some sleep time - want to make sure we're not introducing any flakiness

File: kong/resty/dns/client.lua (hunk 23 above)
Comment (dndx) [reply kong/resty/dns/client.lua:656]: Please see the comment above, without the +1 there is a risk semaphore might timeout before resolve actually happened.

Comment (chobits) [reply kong/resty/dns/client.lua:656]: There are two scenerio where waiting time or timeout
 1. `options.timeout`: a timeout used for network IO of the DNS query
 2. `resolve_max_wait`: waiting for the completion of the DNS query.

If no +1s, these two timeouts will be similar. Consequently, the second scenario may conclude before the first DNS query retrieves the result.

File: spec/01-unit/21-dns-client/02-client_spec.lua (hunk 24 above)
[the remark this answers is not in the document]
Comment (chobits) [reply spec/01-unit/21-dns-client/02-client_spec.lua:1778]: > Also would it be possible to use some kind of waiter function pattern that would check against a predicate every interval ?

If the waiting of the test case happens under two endpoint of C/S model, like that a client wait for server to complete some tasks, your advice is the best way to do. We need to be especially care of handling the synchronization problem between two different process to prevent the flakiness.

But here we only run the DNS query and wait the result in one process. So if this test case fails, there must be some pressing problem happening in the DNS library.

[the remark this answers is not in the document]
Comment (dndx) [reply spec/01-unit/21-dns-client/02-client_spec.lua:1778]: TL;DR: it is a unit test, the chance of flakiness is much lower because client runs much faster than the entire Gateway

**Review by dndx — APPROVED** [review]

**Comment by dndx** [comment]
Test failure due to: https://github.com/Kong/kong/pull/12017

Need to merge that PR & rebase first before merging.
